Minify HTML & CSS, optimize PNGs

Carlrs opened the pull request:
Based on issue #7 

Saves a few kilobytes per fresh uncached page load, will be more significant if/when we can get JS minification working (I'm hoping to get one of the potential JS minification libraries for Rust working properly, rather than using yet another commandline tool, though I won't rule that out either).

**Checklist**
- [x] Updated CHANGELOG.md describing pertinent changes.
- [ ] Updated README.md with pertinent info (may not always apply).
- [ ] Updated `site` content with pertinent info (may not always apply).
- [x] Squash down commits to one or two logical commits which clearly describe the work you've done. If you don't, then Dodd will 🤓.

simbleau commented:
Really happy to see a PR for this. Thanks! Will review this soon.

thedodd commented on src/pipelines/mod.rs:
```diff
@@ -214,11 +224,31 @@ impl AssetFile {
+                }
+                AssetFileType::Icon => {
+                    oxipng::optimize_from_memory(bytes.as_ref(), &Options::from_preset(PNG_OPTIMIZATION_LEVEL))
+                        .expect(format!("error optimizing PNG {:?}", &self.path).as_str())
```
So, instead of using expect, which will panic, we should avoid panics at all costs. Let's use standard error propagation with `?`. And, as we do in other places in this code base, let's use [anyhow::Context](https://docs.rs/anyhow/latest/anyhow/trait.Context.html).
```suggestion
                        .with_context(|| format!("error optimizing PNG {}", &self.path))?
```

Roughly.

thedodd reviewed (changes requested):
Thanks for tackling this! Great work so far, just a few small items.

hamza1311 commented:
This should also minify the JS files created by wasm-bindgen. One approach to do that is by using [SWC](https://swc.rs/)

thedodd commented:
@hamza1311 yea, I am tempted to say that we might want to tread carefully before we pull in SWC. I think it will be awesome to do so, however at that point, we will have everything needed to do full JS/TS transpilation. We have to be fairly measured about whether we really want to push Trunk into that space. 

I am fairly biased towards allowing folks to use hooks or something similar for integration with other tools. As soon as we cross that threshold, we take on the full yoke of the JS ecosystem.

dnaka91 commented:
@hamza1311 @thedodd Despite its name, the `minify-html` crate already includes a JS minifier. Why don't we use that on the wasm-bindgen output?

We definitely should draw a line somewhere though, as thedodd mentioned. For example, for projects that build the whole application in Rust, it's great to have trunk minify everything, as you don't have to tap into any extra tools from the JS ecosystem. But for projects that only use wasm partially or mix and match with JS, it's in my opinion best to leave further bundling to another tool.

Carlrs commented:
Unfortunately, the JS minifier in that crate can't correctly parse the wasm-bindgen output. Incidentally, it's the one I intended to use, so I already created an issue upstream in the parser.

 
 

Sent: Thursday, October 27, 2022 at 12:26 PM
From: "Dominik Nakamura" ***@***.***>
To: "thedodd/trunk" ***@***.***>
Cc: "Carlrs" ***@***.***>, "Author" ***@***.***>
Subject: Re: [thedodd/trunk] Minify HTML & CSS, optimize PNGs (PR #449)


 

@hamza1311 @thedodd Despite its name, the minify-html crate already includes a JS minifier. Why don't we use that on the wasm-bindgen output?

We definitely should draw a line somewhere though, as thedodd mentioned. For example, for projects that build the whole application in Rust, it's great to have trunk minify everything, as you don't have to tap into any extra tools from the JS ecosystem. But for projects that only use wasm partially or mix and match with JS, it's in my opinion best to leave further bundling to another tool.

—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you authored the thread.Message ID: ***@***.***>

hamza1311 commented:
Oh, that's too bad. Can you link to the issue you created here?

dnaka91 commented:
This one, I believe: https://github.com/wilsonzlin/parse-js/issues/3

The `minify-js` crate is relatively new, so maybe best to wait a bit more, before we use it, or we might print a warning and continue on, in case it fails to minify the content. Before they use `esbuild` which pulls in a whole Go project, so I'm really happy they started building a pure Rust implementation instead. Just needs a bit more battle-testing to become more stable over various inputs.

Carlrs commented:
Upstream maintainer for minify-html/minify-js added a fix for parsing import.meta so I'll see about adding JS minification to this PR later today.

thedodd commented:
Just checking in to see if this PR is again ready for review? Maybe a rebase is a good idea too.

Carlrs commented:
At one point I was having issues with it, some with the crates being used (and some of it user error on my behalf), so I decided to let it rest a bit and let the minify-js crate mature a bit, as it was in fairly active development.

I think now is a good time to take a new look at it and test it with new and improved dependencies. I'll try to get around to it in the next few days.

xfbs commented:
@Carlrs I would like to see this feature merged as well. If there is anything I can do to help, feel free to ping me. Would love to help any way I can.

Carlrs commented:
Rebased and fixed the issues I'd had previously (using minify-js in wrong TopLevelMode + not checking for image type for before running oxipng). Let me know how it looks!

MolotovCherry commented:
Do you plan to keep the closing tags, as well as minify js/css in the html?
```rs
let mut minify_cfg = Cfg::spec_compliant();
minify_cfg.minify_css = true;
minify_cfg.minify_js = true;
minify_cfg.keep_closing_tags = true;
```

MolotovCherry commented on src/pipelines/mod.rs:
```diff
@@ -225,11 +241,40 @@ impl AssetFile {
 
+        bytes = if minify {
+            match file_type {
+                AssetFileType::Css => minify_html::minify(&bytes, &Cfg::spec_compliant()),
```
Here, `Cfg::spec_compliant()` returns a `minify_css` default of `false`. Was this intended?

Carlrs replied:
Good catch, it was not intentional

Carlrs commented:
Added your suggested fixes and bumped the minify-html crate version just in case.

@thedodd could we get a review here sometime?

ctron commented:
I cherry-picked this into `trunk-ng`. I will test it for a bit and when there are no problems, it should be released with `0.17.11`.

ctron commented:
This is now released with `trunk-ng` 0.17.11

ctron commented:
This was brought back to trunk with PR https://github.com/trunk-rs/trunk/pull/623 and it should be part of the next release of trunk too.

ctron commented:
This is part of the trunk 0.18.0 release
